pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/exchange/BlockExchange.java

```diff
@@ -58,23 +67,45 @@ public static BlockExchange getExchange(List<SendingMailbox> sendingMailboxes, R
     }
   }
 
-  protected BlockExchange(List<SendingMailbox> sendingMailboxes, BlockSplitter splitter) {
+  protected BlockExchange(OpChainExecutionContext context, List<SendingMailbox> sendingMailboxes,
+      BlockSplitter splitter) {
+    _context = context;
     _sendingMailboxes = sendingMailboxes;
     _splitter = splitter;
+    _context.getMailboxService().submitExchangeRequest(this);
   }
 
-  public void send(TransferableBlock block)
-      throws Exception {
-    if (block.isEndOfStreamBlock()) {
-      for (SendingMailbox sendingMailbox : _sendingMailboxes) {
-        sendBlock(sendingMailbox, block);
+  public void send(TransferableBlock block) {
+    while (!_queue.offer(block)) {
+      if (!_context.yield()) {
+        // failed to send block and context timed out
+        break;
```

Comment:
this is the yielding mechanism --> e.g. we no longer need to back out of the opChain. instead we await here to retry sending once the queue becomes open